File: app/fsharp-2/maui/tests/PuyoPuyo.Tests/IntegrationTests.fs
module PuyoPuyo.Tests.IntegrationTests

open Xunit
open FsUnit.Xunit
open PuyoPuyo.Core.Domain

[<Fact>]
let ``完全なゲームフロー: 着地→消去→重力→連鎖→スコア加算`` () =
    // Arrange: 初期ボード - 下に3つの赤ぷよが縦に配置されている
    let board =
        Board.create 6 13
        |> fun b -> Board.setCell b 1 10 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 11 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)

    // 新しいぷよペア: 赤+青（x=1から開始、rotation=0で上向き）
    let piece = PuyoPair.create 1 1 PuyoColor.Red PuyoColor.Blue 0

    // Act 1: ぷよペアを下まで移動
    let rec moveToBottom (currentPiece: PuyoPair) =
        match GameLogic.tryMovePuyoPair board currentPiece Direction.Down with
        | Some movedPiece -> moveToBottom movedPiece
        | None -> currentPiece

    let landedPiece = moveToBottom piece

    // Act 2: 着地処理
    let boardWithPuyo = Board.fixPuyoPair board landedPiece

    // Assert: ボードに赤ぷよ4つと青ぷよ1つが配置されているはず
    // rotation=0なので、軸ぷよ(Puyo1)がy=9、2つ目のぷよ(Puyo2)がy=8
    Board.getCell boardWithPuyo 1 9 |> should equal (Cell.Filled PuyoColor.Red) // Puyo1
    Board.getCell boardWithPuyo 1 8 |> should equal (Cell.Filled PuyoColor.Blue) // Puyo2
    Board.getCell boardWithPuyo 1 10 |> should equal (Cell.Filled PuyoColor.Red) // 既存
    Board.getCell boardWithPuyo 1 11 |> should equal (Cell.Filled PuyoColor.Red) // 既存
    Board.getCell boardWithPuyo 1 12 |> should equal (Cell.Filled PuyoColor.Red) // 既存

    // Act 3: 連鎖処理（消去と重力を繰り返し適用）、連鎖情報を取得
    let (finalBoard, chainInfo) =
        Board.clearAndApplyGravityRepeatedlyWithInfo boardWithPuyo

    // Assert: 赤ぷよが消えて青ぷよだけが残っているはず
    Board.getCell finalBoard 1 9 |> should equal Cell.Empty
    Board.getCell finalBoard 1 10 |> should equal Cell.Empty
    Board.getCell finalBoard 1 11 |> should equal Cell.Empty

    // Assert: 青ぷよは重力で下に落ちているはず
    Board.getCell finalBoard 1 12 |> should equal (Cell.Filled PuyoColor.Blue)

    // Assert: 連鎖情報を確認
    chainInfo.ChainCount |> should equal 1
    chainInfo.ClearedPuyoCount |> should equal 4

    // Assert: 全消しではない（青ぷよが残っている）
    chainInfo.IsZenkeshi |> should equal false

    // Assert: スコアが計算されている
    let score = Board.calculateScore chainInfo
    score |> should be (greaterThan 0)

[<Fact>]
let ``完全なゲームフロー: 全消しボーナス`` () =
    // Arrange: 初期ボード - 3つの赤ぷよのみ
    let board =
        Board.create 6 13
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 3 12 (Cell.Filled PuyoColor.Red)

    // 新しいぷよペア: 赤2つ
    let piece = PuyoPair.create 1 1 PuyoColor.Red PuyoColor.Red 0

    // Act: 着地まで移動
    let rec moveToBottom (currentPiece: PuyoPair) =
        match GameLogic.tryMovePuyoPair board currentPiece Direction.Down with
        | Some movedPiece -> moveToBottom movedPiece
        | None -> currentPiece

    let landedPiece = moveToBottom piece
    let boardWithPuyo = Board.fixPuyoPair board landedPiece

    // 連鎖処理
    let (finalBoard, chainInfo) =
        Board.clearAndApplyGravityRepeatedlyWithInfo boardWithPuyo

    // Assert: 全て消えているはず
    for y in 0 .. finalBoard.Rows - 1 do
        for x in 0 .. finalBoard.Cols - 1 do
            Board.getCell finalBoard x y |> should equal Cell.Empty

    // Assert: 全消し判定
    chainInfo.IsZenkeshi |> should equal true

    // Assert: 5個のぷよが消えている
    chainInfo.ClearedPuyoCount |> should equal 5

[<Fact>]
let ``完全なゲームフロー: ゲームオーバー判定`` () =
    // Arrange: 初期ボード - 上部までぷよで埋まっている
    let mutable board = Board.create 6 13

    // x=2の列を上部まで埋める（y=0とy=1も埋める）
    for y in 0..12 do
        board <- Board.setCell board 2 y (Cell.Filled PuyoColor.Red)

    // 新しいぷよペア（x=2, y=1が軸、rotation=0なので2つ目はy=0）
    let newPiece = PuyoPair.create 2 1 PuyoColor.Blue PuyoColor.Green 0

    // Act: ゲームオーバー判定
    let isGameOver = GameLogic.checkGameOver board newPiece

    // Assert: ゲームオーバーになるはず
    isGameOver |> should equal true

File: app/csharp-2/maui/PuyoPuyoMAUI.Tests/IntegrationTests.cs
using FluentAssertions;
using PuyoPuyoMAUI.Domain;
using Xunit;

namespace PuyoPuyoMAUI.Tests;

public class IntegrationTests
{
    [Fact]
    public void 完全なゲームフロー_着地から消去_重力_連鎖_スコア加算()
    {
        // 初期ボード: 下に縦に3つの赤ぷよが配置されている
        var board = Board.Create(6, 13)
            .SetCell(2, 10, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red));

        // 新しいぷよペア: 赤と青のペア（縦向き、回転状態0）
        var piece = new PuyoPair(2, 1, PuyoColor.Red, PuyoColor.Blue, 0);

        // 1. ぷよペアを下まで移動
        PuyoPair MoveToBottom(PuyoPair currentPiece)
        {
            var movedPiece = GameLogic.TryMovePuyoPair(board, currentPiece, Direction.Down);
            if (movedPiece != null)
            {
                return MoveToBottom(movedPiece);
            }

            return currentPiece;
        }

        var landedPiece = MoveToBottom(piece);

        // 2. 着地処理
        var boardWithPuyo = board.FixPuyoPair(landedPiece);

        // ボードに赤ぷよが4つ縦に並んでいるはず
        boardWithPuyo.GetCell(2, 9).Color.Should().Be(PuyoColor.Red);  // 新規 Puyo1
        boardWithPuyo.GetCell(2, 10).Color.Should().Be(PuyoColor.Red); // 既存
        boardWithPuyo.GetCell(2, 11).Color.Should().Be(PuyoColor.Red); // 既存
        boardWithPuyo.GetCell(2, 12).Color.Should().Be(PuyoColor.Red); // 既存

        // 青ぷよは上にあるはず
        boardWithPuyo.GetCell(2, 8).Color.Should().Be(PuyoColor.Blue); // 新規 Puyo2

        // 3. 連鎖処理（消去と重力を繰り返し適用）
        var (finalBoard, clearedCounts, isZenkeshi) = boardWithPuyo.ClearAndApplyGravityRepeatedlyWithChainInfo();

        // 4. 赤ぷよが4つ消えているはず
        finalBoard.GetCell(2, 9).IsEmpty.Should().BeTrue();
        finalBoard.GetCell(2, 10).IsEmpty.Should().BeTrue();
        finalBoard.GetCell(2, 11).IsEmpty.Should().BeTrue();
        finalBoard.GetCell(2, 12).IsEmpty.Should().BeFalse(); // 青ぷよが落ちてくる

        // 5. 青ぷよは重力で下に落ちているはず
        finalBoard.GetCell(2, 12).Color.Should().Be(PuyoColor.Blue);

        // 6. 全消しではない（青ぷよが残っている）
        isZenkeshi.Should().BeFalse();

        // 7. 連鎖数は1回
        clearedCounts.Count.Should().Be(1);

        // 8. 消去数は4個
        clearedCounts[0].Should().Be(4);
    }

    [Fact]
    public void 完全なゲームフロー_全消しボーナス()
    {
        // 初期ボード: 3つの赤ぷよのみ
        var board = Board.Create(6, 13)
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(3, 12, Cell.Filled(PuyoColor.Red));

        // 新しいぷよペア: 赤2つ
        var piece = new PuyoPair(1, 1, PuyoColor.Red, PuyoColor.Red, 0);

        // 着地まで移動
        PuyoPair MoveToBottom(PuyoPair currentPiece)
        {
            var movedPiece = GameLogic.TryMovePuyoPair(board, currentPiece, Direction.Down);
            if (movedPiece != null)
            {
                return MoveToBottom(movedPiece);
            }

            return currentPiece;
        }

        var landedPiece = MoveToBottom(piece);
        var boardWithPuyo = board.FixPuyoPair(landedPiece);

        // 連鎖処理
        var (finalBoard, clearedCounts, isZenkeshi) = boardWithPuyo.ClearAndApplyGravityRepeatedlyWithChainInfo();

        // 全て消えているはず
        for (int y = 0; y < finalBoard.Rows; y++)
        {
            for (int x = 0; x < finalBoard.Cols; x++)
            {
                finalBoard.GetCell(x, y).IsEmpty.Should().BeTrue();
            }
        }

        // 全消し判定
        isZenkeshi.Should().BeTrue();

        // 連鎖数は1回
        clearedCounts.Count.Should().Be(1);

        // 消去数は5個（既存3個 + 新規2個）
        clearedCounts[0].Should().Be(5);
    }

    [Fact]
    public void 完全なゲームフロー_ゲームオーバー判定()
    {
        // 初期ボード: 上部までぷよで埋まっている
        var board = Board.Create(6, 13);
        for (int y = 1; y <= 12; y++)
        {
            board = board.SetCell(2, y, Cell.Filled(PuyoColor.Red));
        }

        // 新しいぷよペア
        var newPiece = new PuyoPair(2, 1, PuyoColor.Blue, PuyoColor.Green, 0);

        // ゲームオーバー判定
        var isGameOver = GameLogic.CheckGameOver(board, newPiece);

        // ゲームオーバーになるはず
        isGameOver.Should().BeTrue();
    }
}
